### DISCUSSION TITLE
Swift package extraction does not work when dynamically defining dependencies

### DISCUSSION THREAD

**Discussion by G00fY2** [renovatebot/renovate#33089] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket Server

### Please tell us more about your question or problem

We define dependencies in our `Package.swift` file dynamically like this:

```swift
let package = Package(
    name: "PackageName",
    dependencies: makePackageDependencies()
)

func makePackageDependencies() -> [PackageDescription.Package.Dependency] {
    var dependencies: [PackageDescription.Package.Dependency] = [
        .package(url: "https://github.com/Alamofire/Alamofire.git", exact: "5.10.1"),
    ]

    if PackageEnvironment.get() == .debug {
        dependencies += [
            .package(url: "https://github.com/FLEXTool/FLEX", exact: "5.22.10")
        ]
    }

    return dependencies
}
```

Seems like this is not supported by Renovate (since no dependencies are extracted currently).

Should renovate maybe just scan all occurrences of `.package(` and not strictly assume the "default" package file hierarchy?

I think this is related to https://github.com/renovatebot/renovate/discussions/31980?

### Logs (if relevant)

<details><summary>Logs</summary>

```

[2024-12-13T11:11:57.821Z] DEBUG: Matched 1 file(s) for manager mise: .mise.toml (repository=MA/ios-client)
[2024-12-13T11:11:57.821Z] DEBUG: Matched 1 file(s) for manager swift: Package.swift (repository=MA/ios-client)
[2024-12-13T11:11:57.821Z] DEBUG: manager extract durations (ms) (repository=MA/ios-client)
[2024-12-13T11:11:57.821Z]        "managers": {"git-submodules": 0, "mise": 3, "swift": 4}
[2024-12-13T11:11:57.821Z] DEBUG: Found mise package files (repository=MA/ios-client)
[2024-12-13T11:11:57.821Z] DEBUG: Found 1 package file(s) (repository=MA/ios-client)
[2024-12-13T11:11:57.821Z]  INFO: Dependency extraction complete (repository=MA/ios-client, baseBranch=develop)
[2024-12-13T11:11:57.821Z]        "stats": {
[2024-12-13T11:11:57.821Z]          "managers": {"mise": {"fileCount": 1, "depCount": 4}},
[2024-12-13T11:11:57.821Z]          "total": {"fileCount": 1, "depCount": 4}
[2024-12-13T11:11:57.821Z]        }

```

</details>

**Comment by rarkins** [comment]
This syntax is not supported. In theory there's an infinite number of ways which dynamic dependency definitions could be done, including with string interpolation. In practice maybe it can be supported if we stick to a "dumb" approach of searching for `.package`.

@viceice wdyt?

**Reply by G00fY2** [reply]
We can probably workaround this by conditionally filter or append packages to the list of the default syntax.
But maybe this limitation could be documented in the renovate docs somewhere.

**Reply by rarkins** [reply]
PRs welcome to here: https://github.com/renovatebot/renovate/blob/main/docs/usage/nuget.md

**Reply by viceice** [reply]
can work. I'm not familiar with that manager and those files

**Reply by mflknr** [reply]
Is looking into the 'Package.resolved' file worth a consideration? It is generated by SPM - not sure if thats a problem in on itself - and standardised.

And it would also enable renovate to not only work for Swift Packages but also projects that uses Xcodes SPM integration AFAIK.

**Reply by mflknr** [reply]
Ah, nvm. Wasn't sure how renovate operates but using 'Package.resolved' seems not feasable in any way whatsoever.

**Reply by rarkins** [reply]
We don't run binaries / third party tools as part of the extract process

**Reply by gralamin** [reply]
This ended up being my problem as well. I'm confused as to why the `Package.resolved` is not used? It is a `json` file and is basically the same as node's `package-lock.json`. Can someone elaborate on the reason why that isn't viable?

**Reply by rarkins** [reply]
There's no problem looking at the resolved file if that can give us the info

**Reply by gralamin** [reply]
After thinking more on this, it should be enough to figure out an update is required, but it will remain tricky to figure out how to update the dependencies. The workaround mentioned in the second thread is probably the best solution for now.

**Comment by rarkins** [comment]
A custom manager looking for the `.package()` pattern could be a good workaround for now

**Comment by maartene** [comment]
Just upvoted this issue. As dynamic elements in `Package.swift` files are very common.

Maybe checking the `Package.resolved` file instead could be an option? (Ah, I see this was already suggested :) )
